test/cases: fix users stage

kingsleyzissou opened the pull request:
Since the `/etc/shadow` file contains a timestamp we need to add a `null` value rather than a `sha256` hash to tell the diff tool to ignore these fields. The issue is that the timestamp will always be different meaning the tests will pass for a day, but then fail after that.

mvo5 commented on tools/tree-diff:
```diff
@@ -64,7 +64,10 @@ def content_diff(name, dir_fd1, dir_fd2, path, differences):
         if hash1 != hash2:
             props = differences.setdefault(path, {})
-            props["content"] = [hash1, hash2]
+            if name == "shadow" or name == "shadow-":
```
Is this part of the diff still needed when the content is set to "null", "null" in the above "diff.json"?

thozza replied:
AFAIK, no....

kingsleyzissou replied:
Just so I'm on the same page, you mean leave the section out entirely?

thozza replied:
it should be enough to set the hash values to `null` in the `diff.json` of the respective stage test. There shouldn't be any need to modify `tools/tree-diff`.

kingsleyzissou replied:
Ah right. Will fix

thozza commented:
@kingsleyzissou basically just revert to https://github.com/osbuild/osbuild/commit/ad265a519d3f694d313192b1aba97c6cbcec3e3a#diff-2d52867474e88bf90ffccbf84880460deb017f04f971591cfd2a22c6ec1d05e4L81-L92 without any additional changes...

kingsleyzissou commented:
> @kingsleyzissou basically just revert to [ad265a5#diff-2d52867474e88bf90ffccbf84880460deb017f04f971591cfd2a22c6ec1d05e4L81-L92](https://github.com/osbuild/osbuild/commit/ad265a519d3f694d313192b1aba97c6cbcec3e3a#diff-2d52867474e88bf90ffccbf84880460deb017f04f971591cfd2a22c6ec1d05e4L81-L92) without any additional changes...

Updated. Yeah I had added the changes so the `tools/gen-stage-test-diff` would generate the correct diff.

thozza commented:
> Updated. Yeah I had added the changes so the `tools/gen-stage-test-diff` would generate the correct diff.

Ah, I didn't get that, sorry. I have never used `tools/tree-diff` for generating the test diff :innocent: I'm not opposed to modifying the tool that generates the diff, so that we don't have to modify the diff manually...

kingsleyzissou commented:
> Ah, I didn't get that, sorry. I have never used tools/tree-diff

That's okay. I don't think it was a very complete list though, since there are other stages that would have the same issue. Might need to think about it a bit more.

mvo5 commented:
> > Ah, I didn't get that, sorry. I have never used tools/tree-diff
> 
> That's okay. I don't think it was a very complete list though, since there are other stages that would have the same issue. Might need to think about it a bit more.

Aha, sorry! My comment was born of ignorance then, I assumed this part of the diff was to fix the failure, I did not realize that that was not the purpose. Maybe worth doing a second PR with the fix in the gen-diff-tool? The alternative would be that we run our tests with `faketime` but maybe the extra work is not worth it (and certain tools written in go or rust will not work with it as it does LD_PRELOAD tricks that will not work on those more modern languages.

kingsleyzissou commented:
> Aha, sorry! My comment was born of ignorance then, I assumed this part of the diff was to fix the failure, I did not realize that that was not the purpose. Maybe worth doing a second PR with the fix in the gen-diff-tool? The alternative would be that we run our tests with `faketime` but maybe the extra work is not worth it (and certain tools written in go or rust will not work with it as it does LD_PRELOAD tricks that will not work on those more modern languages.

No worries :) I just created https://github.com/osbuild/osbuild/pull/1802, it's a pretty naive fix though
